### ISSUE TITLE
Feature Request - get tags of (Hibernated) WebSocket

### ISSUE THREAD

**Issue by DaniFoldi** [cloudflare/workerd#958]
Hi :wave:,

We began adopting the new Hibernation API in projects, and I have an idea for an improvement of the APIs it has. Currently, the `serializeAttachment` and `deserializeAttachment` methods are provided to persist data that needs to be read back. Additionally, the `tags` attribute of `state.acceptWebSocket` can be used for retrieving the WebSockets.

And this is where I see a gap in the API - there is no way to retrieve _other tags_ of a WebSocket instance, in a listener or after using `getWebSockets`. Besides other use cases, it would be useful for libraries which want to label WebSockets somehow (as an example see https://github.com/evanderkoogh/otel-cf-workers/issues/21#issuecomment-1661325331).

Yes, the library could also use the attachment to save its data, however I see the 2048-byte attachment as a much more "valuable" space for developers, that some people (me for sure) are likely to calculate the exact bytes required to stuff in as much data as possible. On a similar note, I haven't yet found a use case which requires 10 tags on a WebSocket for retrieval later, so if one of these tags could be used by libraries (note: the only missing bit here is the retrieval of this tag from the WS instance) like otel-cf-workers to save their data transparently to the user, the 2048 bytes that the _user_ is working with wouldn't have to be edited by _libraries_.

**Comment by DaniFoldi** [comment]
I'll propose two possible solutions here, as the first one would be adequate for most usecases, but it feels natural to expand to the full suite.

At a minimum I'm thinking of a `WebSocket.getTags(): string[] | null` method that returns all tags provided when the WS was handed to the runtime for hibernation. This way, I could retrieve all other tags of a WebSocket when it wakes up, for example if the tags are `user:1234` and `room:ABC`, I can get all the sockets from a room (say for broadcasting a message) by `getWebSockets('room:ABC')` - but maybe I want to know where they are connected from, so I'd look for the `country:<something>` tag - hard to know in advance, but useful information.

Expanding on the idea, an update method makes sense to me where I can edit the tags of a WebSocket. I don't currently have a use case for this, but it completes the API surface for these tags, same as the serialized attachment.

I know that some of these can be solved with the attachment system, however the 10 tags * 256 bytes are also currently allowed, so this change wouldn't require any new storage per WS.

**Comment by MellowYarker** [comment]
### Getting Tags
Would `state.getTags(WebSocket)` be alright? We don't want to stray too far from the official spec, so we'd like to avoid extending the `WebSocket` object if we can.

I think we can support this but it may be tricky to implement. A `HibernatableWebSocket` wraps a regular `jsg::WebSocket`, so we'd need a way to access the wrapper (where the tags are) from the inner object. This may sound straightforward (give it a ref!), however there are situations where the `HibernatableWebSocket` can be dtor'd before the `jsg::WebSocket` (Ex. in a `webSocketClose/Error()` handler). In that case, you could see non-deterministic behavior where you can't get the tags back in those handlers.

One way we could get around that is to copy all the tags into the `jsg::WebSocket` object, but that would increase the memory usage of the DO in the isolate. As you can imagine, if there are a lot of active websockets in an isolate, each with a couple of longer tags, you could eat a good chunk of your memory (worst case 32k ws, 10 tags each 256 chars: almost 84MB!). Maybe this isn't a real concern though... since we could limit this to when we know the `HibernatableWebSocket` wrapper object is going away -- then it's a once-per-connection increase in memory right before the `WebSocket` is GC'd.

**tl;dr**:
- `WebSocket.getTags()` -> `state.getTags(WebSocket)`
- expect slight increase in memory when `webSocketClose/webSocketError()` run (at most 10 tags * 256 chars ~2560 bytes per ws, maybe scary when processing many simultaneous disconnects?)

---
### Updating Tags
I would slightly push back against providing a way to update tags until we know there's a genuine need to do so. Our mechanism for storing tags (and associating websockets with them) is fairly complicated, as we optimized for accessing websockets by tag, removing disconnected websockets efficiently, and decreasing our memory footprint (both inside and outside of the isolate).

As far as an API for this is concerned, would you prefer having a way to add/remove tags (within the existing limits), rather than just updating existing ones?

**Comment by DaniFoldi** [comment]
Hi Milan :wave:,

Thanks for considering my feature request. I don't see any downside of having the API flipped - code using `state.getTags(WebSocket)` would look exactly the same in the real world, in my opinion.

I hope I understand the issue you're describing with the non-deterministic behavior, however it's been a while since I've written c++, so even though I've read the relevant runtime code, the the context I have about the memory increase is probably rather limited. Even so, I think that the memory footprint increase is small enough to justify the new feature. My only concern around this is the runtime restarting/stopping, when all hibernated websockets are closed, are the close handlers on each connected socket called at the same time, thus potentially reaching the 84MB increase?

As for the update method, as I mentioned above I don't have a real use case yet, so I understand the reason for not implementing that for now. Since you asked, I was thinking that the simplest (although maybe this is not true given your explanation) to implement is a `setTags`, that accepts a new array of tags, replacing all the existing ones.

**Comment by MellowYarker** [comment]
We're tracking this internally, haven't had time to look into the restart issue (not sure if it even is an issue). Will update here when I have some time to dig through the code/run some tests.

**Comment by DaniFoldi** [comment]
Thank you @MellowYarker for implementing this!

I'm going to close this issue since the PR has now been merged (yayy :tada: ) and released - eagerly waiting for this to appear on Workers.
